### ISSUE TITLE
c10d communication failure on k8s cluster

### ISSUE THREAD

**Issue by streamnsight** [pytorch/torchx#541]
## 🐛 Bug
cross-referencing the issue posted here:

https://github.com/pytorch/pytorch/issues/80772

which is about race condition between rank 0 pod registering with DNS and c10d service trying to look it up.

## To Reproduce

1. run the code with `torchx run --scheduler kubernetes dist.ddp -j 8x1 --script cifar_dist.py`
2. see the pods scheduled and then erroring

## Code:
```
import os
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
from torch.utils.data.distributed import DistributedSampler
from torch.utils.data import DataLoader

from torchvision.models import resnet18
from torchvision.datasets import CIFAR10
import torchvision.transforms as transforms

from torch.nn.parallel import DistributedDataParallel as DDP
import random
import numpy as np
from time import sleep

model_dir = '/model'
model_filename = 'mymodel'
batch_size = 128


def set_random_seeds(seed=0):

    torch.manual_seed(seed)
    # torch.backends.cudnn.deterministic = True
    # torch.backends.cudnn.benchmark = False
    np.random.seed(seed)
    random.seed(seed)


def evaluate(model, device, test_loader):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for data in test_loader:
            images, labels = data[0].to(device), data[1].to(device)
            outputs = model(images)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()
    accuracy = correct / total
    return accuracy


dist.init_process_group("gloo")
rank = dist.get_rank()
world_size = dist.get_world_size()

if torch.cuda.is_available():
    device = torch.cuda.device(f'cuda:{rank}')
else:
    device = torch.device('cpu')


print(f"Running basic DDP example on rank {rank} of {world_size}")

if rank == 0:
    model_filepath = os.path.join(model_dir, model_filename)
    if not os.path.exists(model_dir):
        os.makedirs(model_dir)

set_random_seeds()

print('create model...')
# create model
model = resnet18(pretrained=False)
if torch.cuda.is_available():
    torch.cuda.set_device(rank)
    device = torch.cuda.device(f'cuda:{rank}')
    ddp_model = DDP(model, device_ids=[rank], output_device=rank)
else:
    device = torch.device('cpu')
    ddp_model = DDP(model)

print('prepare dataset...')
# Prepare dataset and dataloader
transform = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
])

train_set = CIFAR10(root="data", train=True, download=True, transform=transform) 
test_set = CIFAR10(root="data", train=False, download=True, transform=transform)

train_sampler = DistributedSampler(dataset=train_set)

train_loader = DataLoader(dataset=train_set, batch_size=batch_size, sampler=train_sampler, num_workers=0)
# Test loader does not have to follow distributed sampling strategy
test_loader = DataLoader(dataset=test_set, batch_size=128, shuffle=False, num_workers=0)

print('setup model...')

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(ddp_model.parameters(), lr=0.1, momentum=0.9, weight_decay=1e-5)

for epoch in range(1000):
    print(f'Epoch {epoch}')
    if epoch % 10 == 0:
        if rank == 0:
            accuracy = evaluate(model=ddp_model, device=device, test_loader=test_loader)
            torch.save(ddp_model.state_dict(), model_filepath)
            print("-" * 75)
            print(f"Epoch: {epoch}, Accuracy: {accuracy}")
            print("-" * 75)
        else:
            print("-" * 75)
            print(f"Epoch: {epoch}")
            print("-" * 75)

    ddp_model.train()

    i = 0
    for data in train_loader:
        print(i)
        i += 1
        inputs, labels = data[0].to(device), data[1].to(device)
        optimizer.zero_grad()
        outputs = ddp_model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
    
```

## Expected behavior

I expect pods to connect to c10d properly and run the code.

## Environment

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.17-2136.306.1.3.el8uek.x86_64-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] botorch==0.6.0
[pip3] gpytorch==1.6.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] pytorch-lightning==1.5.10
[pip3] torch==1.11.0
[pip3] torch-model-archiver==0.6.0
[pip3] torchelastic==0.2.2
[pip3] torchmetrics==0.9.1
[pip3] torchserve==0.6.0
[pip3] torchtext==0.12.0
[pip3] torchvision==0.12.0
[pip3] torchx==0.2.0
[conda] blas                      1.0                         mkl  
[conda] botorch                   0.6.0                    pypi_0    pypi
[conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] gpytorch                  1.6.0                    pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.21.2           py38h20f2e39_0  
[conda] numpy-base                1.21.2           py38h79a1101_0  
[conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-lightning         1.5.10                   pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-model-archiver      0.6.0                    pypi_0    pypi
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchmetrics              0.9.1                    pypi_0    pypi
[conda] torchserve                0.6.0                    pypi_0    pypi
[conda] torchtext                 0.12.0                     py38    pytorch
[conda] torchvision               0.12.0               py38_cu113    pytorch
[conda] torchx                    0.2.0                    pypi_0    pypi
```

 - Execution environment: Oracle Cloud OKE cluster k8s version 1.23.4
 - dns: CoreDNS-1.8.6 linux/amd64, go1.16.7 BoringCrypto, 69a006c9f1a

**Comment by streamnsight** [comment]
since this is a k8s specific issue, I tried a few things, and adding an init container that waits for DNS resolution seems to fix the problem.

the code in `kubernetes_scheduler.py`  looks like:
```
    container = ...
    
    rank_0_name = re.sub(r'-\d', '_0', name).upper()
    rank_0_hostname = f"$(VC_{rank_0_name}_HOSTS)"
    
    init_container = V1Container(
        command=['/bin/sh'],
        args=['-c', f"echo {rank_0_hostname} && while ! nslookup {rank_0_hostname}; do sleep 1; done"],
        image='busybox:1.28',
        name=name + '-init',
        env=[
            V1EnvVar(
                name=name,
                value=value,
            )
            for name, value in role.env.items()
        ],
        resources=resources,
        security_context=security_context,
    )


    return V1Pod(
        spec=V1PodSpec(
            containers=[container],
            init_containers=[init_container],
            ...
```

**Comment by atinsood** [comment]
@streamnsight do you happen to have the logs handy. also I think there might be an issue with whether you have the ports open or not by default

**Comment by streamnsight** [comment]
@atinsood 
Sorry I didn't repost the logs here, as they are in the linked issue in the pytorch project
https://github.com/pytorch/pytorch/issues/80772

It does not 'always' fail, and when it doesn't, it works fine. So this is definitely a race condition issue with the rank0 pod not being resolved by the DNS service. The initContainer code I show above fixes the problem by waiting for rank0 pod DNS to resolve before starting the other rank pods.

**Comment by atinsood** [comment]
@streamnsight I added some comments to you issue on the pytorch issue, the other thing I realized is that in torchx the store is always hardcoded to be c10d https://github.com/pytorch/torchx/blob/main/torchx/components/dist.py#L132 . you can always provide a new config, but if you are using `dist.ddp` its using c10d by default

**Comment by atinsood** [comment]
I am not 100% sure what this means https://github.com/pytorch/torchx/blob/main/torchx/components/dist.py#L179. I have a feeling that this eventually maps down to container port in kube, but I am not sure and if that is the case if I end up using a port other than 29500 by specifying master_port would I need to change this.

**Comment by streamnsight** [comment]
@atinsood 
OK, I'm not sure what all your suggestions do for me: torchx is the tool that creates he config, and passes the hostname of the rank0 pod as the host to resolve. It is a dynamic config since the name itself is defined by torchx. 
The hostname resolution is the problem. Or rather, the resolution of the hostname in the rank 1+ pods is the problem when the rank0 pod is starting. Hence my solution to add an initContainer to first wait for the rank0 pod to resolve, and then other pods work fine.
I tested this; it work.

Now, using etcd will be the same problem if torchx is creating the dynamic config and pods rank 1+ are trying to resolve pod rank0 before it is registered.

So either the pytorch dist.ddp has a way to wait for the rank0 to be resolved (i.e. fix this issue in pytorch.dist by adding a way to retry name resolutio) or since it is a kubernetes specific problem, provide a kubernetes specific solution like I did (i.e. add an initContainer that does the waiting for DNS resolution)

**Comment by atinsood** [comment]
@streamnsight I think I might have misunderstood the question. it seems like in your case the dns resolution is a problem because you think the pods are not fully up. and you are thinking of fixing that with the init container. that's fine. I am not sure how torchx works with kubernetes. but it seems like they are creating volcano jobs
also from their docs:


```
TorchX kubernetes scheduler depends on volcano and requires etcd intalled for distributed job execution.

Install volcano 1.4.0 version

.. code:: bash

    kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/v1.4.0/installer/volcano-development.yaml

TorchX uses `torch.distributed.run <https://pytorch.org/docs/stable/elastic/run.html>`_ to run distributed training.
This requires the installation of etcd service on your kubernetes cluster:

.. code:: bash

    kubectl apply -f https://github.com/pytorch/torchx/blob/main/resources/etcd.yaml


Learn more about running distributed trainers :py:mod:`torchx.components.dist`

"""```

**Comment by streamnsight** [comment]
yes, torchx creates volcano jobs, with dynamic hostname defined. and yes the issue is the race condition when trying to resolve pod with rank0, if pod with rank 1+ starts before rank 0 is registered in DNS.

I provided a solution which works. See code above. Unfortunatetly, I can't create a PR because I can't sign the CLA, but if someone would be so kind to create one with that change, that would resolve this issue. The init container solution is not something you can add as a config unfortunately.

**Comment by d4l3k** [comment]
@streamnsight can you test out PyTorch 1.12.0? In PyTorch 1.11.0 DNS lookup wouldn't be retried but under 1.12.0 dns failure is a retryable error.

https://github.com/pytorch/pytorch/pull/74641 fixes the issue

**Comment by streamnsight** [comment]
@d4l3k You reference the PR I mentioned, but as far as I can see this PR was closed and NOT merged. So are you sure this is fixed?

**Comment by d4l3k** [comment]
@streamnsight it's in 1.12 https://github.com/pytorch/pytorch/commit/5b915e844caf307afacc6aa6fa21fdcda0ba967d

![2022-07-12-162449_881x153_scrot](https://user-images.githubusercontent.com/909104/178615027-f9302284-5c55-4874-a8ab-15e37d506bda.png)

PyTorch uses a custom bot to merge code and not via the GitHub merge button

**Comment by streamnsight** [comment]
@d4l3k So I tried with a custom image upgrading torch to 1.12, and it does seem to work.
It failed at first because the image pull took longer on one of my nodes where the rank pod ended up being scheduled, and the c10d timed-out, but in next try with the image layers being mostly there already, it worked fine.

the `Dockerfile.torchx` I used was:

```Dockerfile
FROM ghcr.io/pytorch/torchx:0.2.0
RUN pip install -U torch
COPY cifar_dist.py .
```

**Comment by streamnsight** [comment]
Closing this since I got a fix. Hopefully torch 1.12 is in the next release so there is no need for the custom image.
